src/silicone/stats.py

```diff
@@ -103,37 +104,122 @@ def rolling_window_find_quantiles(
+    smoothing : float or string
+        By default, no smoothing is done on the distribution. If an argument is used,
+        it is fed into scipy.stats.gaussian_kde. If a float is input, we
+        fit a Gaussian kernel density estimator with that width to the points and return
+        the quantiles of that distribution. If a string is used, it must be either
+        "scott" or "silverman", after those two methods of determining the best kernel
+        bandwidth. 
+    weighting : None or Series
+        If a series, must have the same indices as distribution, giving the
+        relative weights of each point.
+    to_quantile : Bool
+        If true, we return the quantiles of the data in points_to_quant. If False, we
+        instead treat points_to_quant as the quantiles themselves (they must all be 0-1)
+        and return the values in distribution that occur at these quantiles.
 
     Returns
     -------
     :obj:`np.ndarray`
         an array of width 1 and the same length as points_to_quant, containing the
-        quantiles of these points in order.
+        quantiles of these points in order. Or, if to_quantile is false, containing the
+        values corresponding to the quantiles points_to_quant.
     """
-    distribution = distribution[~np.isnan(distribution)]
+    if not to_quantile:
+        assert all(
+            [i <= 1 and i >= 0 for i in points_to_quant]
+        ), "Quantiles are not defined outside the range 0-1"
+    if isinstance(weighting, type(None)):
+        weighting = pd.Series(
+            [1 for i in range(len(distribution))], index=distribution.index
+        )
+    else:
+
+        assert type(weighting) == pd.Series, "The weighting variable should be a Series"
+        assert len(weighting) == len(distribution), (
+            "There must be the same number of weights as entries in the database. "
+            "We have len {} in weights and len {} in distribution.".format(
+                len(weighting), len(distribution)
+            )
+        )
+        assert np.array_equal(
+            weighting.index.sort_values(), distribution.index.sort_values()
+        ), (
+            "The entries in the weighting Series are not clearly aligned with the "
+            "distribution Series. We have \n {} \n in the weighting and \n {} \n in "
+            "the distribution ".format(weighting.index, distribution.index)
+        )
+
+    if any(np.isnan(distribution)):
+        distribution = distribution[~np.isnan(distribution)]
+        weighting = weighting[distribution.index]
     len_lead_not_nan = len(distribution)
     if len_lead_not_nan == 1:
-        # If there is only a single value then quantile is not defined and we return nan
-        return np.array([np.nan for a in range(len(points_to_quant))])
+        # If there is only a single value then quantile is not defined
+        if to_quantile:
+            return np.array([np.nan for a in range(len(points_to_quant))])
```

Comment:
```suggestion
            return np.nan * np.zeros_like(points_to_quant)
```

Or something like that will be quicker

Reply:
ping